### PR TITLE
Get pose from apriltags

### PR DESCRIPTION
**Pull request by Tyson453**

### CODE REVIEW

File: robotbase/src/main/java/com/team2357/frc2023/networktables/AprilTagPose.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    private static AprilTagPose m_instance;
+
+    public static AprilTagPose getInstance() {
+        if (m_instance == null) {
+            m_instance = new AprilTagPose();
+        }
+        return m_instance;
+    }
+
+    private NetworkTable m_table = NetworkTableInstance.getDefault().getTable("apriltag");
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/frc2023/networktables/AprilTagPose.java:25]: Use a constant here please, we don't know for sure what this will be.

File: robotbase/src/main/java/com/team2357/frc2023/networktables/AprilTagPose.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+    private StringSubscriber m_poseSub;
+
+    private JSONParser m_parser = new JSONParser();
+
+    private AprilTagPose() {
+        NetworkTableInstance inst = NetworkTableInstance.getDefault();
+
+        m_poseSub = m_table.getStringTopic("pose").subscribe("");
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/frc2023/networktables/AprilTagPose.java:34]: Same here, let's use a constant.

File: robotbase/src/main/java/com/team2357/frc2023/networktables/AprilTagPose.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+        m_instance = this;
+    }
+
+    public Pose2d getPose() {
+        return toPose2d(m_poseSub.get());
+    }
+
+    public Pose2d toPose2d(String jsonString) {
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/frc2023/networktables/AprilTagPose.java:43]: This is fine for now, but we may want to break `toPose2d`, `getTranslation` and `getRotation` out into a utility file. I could see us needing these functions outside of the scope of getting a JSON blob from a network table topic. Additionally, you would not have to expose the process you are using to get the pose just for testing purposes.

File: robotbase/src/test/java/com/team2357/frc2023/networktables/AprilTagPoseTest.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
+package com.team2357.frc2023.networktables;
```
Comment (NolanCampbell2357) [inline robotbase/src/test/java/com/team2357/frc2023/networktables/AprilTagPoseTest.java:1]: These test are amazing! I would suggest you add one to test an empty json object to make sure exceptions are not thrown. There might be a case where we lose the cameras, or the cameras do not see an apriltag, and may just return an empty object. We need to make sure the robotcode does not crash when that happens.

File: robotbase/src/main/java/com/team2357/frc2023/networktables/AprilTagPose.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+                .get("pose")).get("translation");
+        translationX = (double) translation.get("x");
+        translationY = (double) translation.get("y");
+
+        return new Translation2d(translationX, translationY);
+    }
+
+    public Rotation2d getRotation(JSONObject obj, int tagIdx) {
+        JSONObject quaternion = (JSONObject) ((JSONObject) ((JSONObject) ((JSONObject) ((JSONArray) (obj.get("tags"))).get(tagIdx)).get("pose")).get("rotation")).get("quaternion");
+        double w = (double) quaternion.get("W"), x = (double) quaternion.get("X"), y = (double) quaternion.get("Y"), z = (double) quaternion.get("Z");
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/frc2023/networktables/AprilTagPose.java:70]: It's typically bad practice to initialize multiple variables on the same line. It's one of those things you can do, but leads to messy code.

**Review by NolanCampbell2357 — CHANGES_REQUESTED** [review]
Looks really good. Just a few things.

**Review by NolanCampbell2357 — APPROVED** [review]
Well done.
